### PR TITLE
C++ API parity: AdaptiveMaxPool1d

### PR DESCRIPTION
**Pull request by pbelevich**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26915 C++ API parity: MaxUnpool2d
* #26896 C++ API parity: MaxUnpool1d
* #26819 C++ API parity: AdaptiveAvgPool3d
* #26818 C++ API parity: AdaptiveAvgPool2d
* #26808 C++ API parity: AdaptiveAvgPool1d
* #26775 C++ API parity: AdaptiveMaxPool3d
* #26772 C++ API parity: AdaptiveMaxPool2d
* **#26755 C++ API parity: AdaptiveMaxPool1d**

Differential Revision: [D17627828](https://our.internmc.facebook.com/intern/diff/D17627828)

### CODE REVIEW

**Review by yf225 — COMMENTED** [review]
Thanks @pbelevich ! I left some comments.

File: test/cpp_api_parity/parity-tracker.md [hunk 1]
```diff
@@ -29,7 +29,7 @@ torch.nn.AvgPool3d|Yes|No
 torch.nn.FractionalMaxPool2d|No|No
 torch.nn.LPPool1d|No|No
 torch.nn.LPPool2d|No|No
-torch.nn.AdaptiveMaxPool1d|No|No
+torch.nn.AdaptiveMaxPool1d|Yes|No
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply test/cpp_api_parity/parity-tracker.md]: reverted

File: torch/csrc/api/include/torch/nn/options/pooling.h [hunk 2]
```diff
@@ -79,5 +79,20 @@ using MaxPool2dOptions = MaxPoolOptions<2>;
...
+struct AdaptiveMaxPoolOptions {
+  AdaptiveMaxPoolOptions(ExpandingArray<D> output_size)
+      : output_size_(output_size) {}
+
+  /// the target output size
+  TORCH_ARG(ExpandingArray<D>, output_size);
+};
+
+/// `AdaptiveMaxPoolOptions` specialized for 1-D maxpool.
+using AdaptiveMaxPool1dOptions = AdaptiveMaxPoolOptions<1>;
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply torch/csrc/api/include/torch/nn/options/pooling.h]: fixed

File: torch/csrc/api/include/torch/nn/functional/pooling.h [hunk 3]
```diff
@@ -70,6 +70,18 @@ inline Tensor max_pool3d(const Tensor& input, const MaxPool3dOptions& options) {
...
+
+inline Tensor adaptive_max_pool1d(const Tensor& input,
+  const AdaptiveMaxPool1dOptions& options) {
+   return std::get<0>(torch::adaptive_max_pool1d(input, options.output_size()));
+}
+
+inline std::tuple<Tensor, Tensor> adaptive_max_pool1d_with_indices(
+  const Tensor& input, const AdaptiveMaxPool1dOptions& options) {
+   return torch::adaptive_max_pool1d(input, options.output_size());
+}
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply torch/csrc/api/include/torch/nn/functional/pooling.h]: addressed

**Review by yf225 — APPROVED** [review]
Looks awesome, thanks @pbelevich !

**Pull request closed without merging**